src/emu/emuopts.cpp

```diff
@@ -213,133 +218,777 @@ const options_entry emu_options::s_option_entries[] =
+		}
+		return result;
+	}
+}
+
+
+//**************************************************************************
 //  EMU OPTIONS
 //**************************************************************************
 
 //-------------------------------------------------
 //  emu_options - constructor
 //-------------------------------------------------
 
-emu_options::emu_options()
-: core_options()
-, m_coin_impulse(0)
-, m_joystick_contradictory(false)
-, m_sleep(true)
-, m_refresh_speed(false)
-, m_ui(UI_CABINET)
+emu_options::emu_options(bool general_only)
+	: m_system(nullptr)
+	, m_coin_impulse(0)
+	, m_joystick_contradictory(false)
+	, m_sleep(true)
+	, m_refresh_speed(false)
+	, m_ui(UI_CABINET)
 {
+	// add entries
+	if (!general_only)
+	{
+		add_entry(std::make_unique<system_name_option_entry>(*this));
+		add_entry(std::make_unique<software_name_option_entry>(*this));
+	}
 	add_entries(emu_options::s_option_entries);
+
+	// adding handlers to keep copies of frequently requested options in member variables
+	set_value_changed_handler(OPTION_COIN_IMPULSE,				[this](const char *value) { m_coin_impulse = int_value(OPTION_COIN_IMPULSE); });
+	set_value_changed_handler(OPTION_JOYSTICK_CONTRADICTORY,	[this](const char *value) { m_joystick_contradictory = bool_value(OPTION_JOYSTICK_CONTRADICTORY); });
+	set_value_changed_handler(OPTION_SLEEP,						[this](const char *value) { m_sleep = bool_value(OPTION_SLEEP); });
+	set_value_changed_handler(OPTION_REFRESHSPEED,				[this](const char *value) { m_sleep = bool_value(OPTION_REFRESHSPEED); });
+	set_value_changed_handler(OPTION_UI, [this](const std::string &value)
+	{
+		if (value == "simple")
+			m_ui = UI_SIMPLE;
+		else
+			m_ui = UI_CABINET;
+	});
+}
+
+
+//-------------------------------------------------
+//  emu_options - destructor
+//-------------------------------------------------
+
+emu_options::~emu_options()
+{
+}
+
+
+//-------------------------------------------------
+//	system_name
+//-------------------------------------------------
+
+const char *emu_options::system_name() const
+{
+	return m_system ? m_system->name : "";
 }
 
 
 //-------------------------------------------------
-//  value_changed - to prevent tagmap
-//    lookups keep copies of frequently requested
-//    options in member variables.
+//  set_system_name - called to set the system
+//	name; will adjust slot/image options as appropriate
 //-------------------------------------------------
 
-void emu_options::value_changed(const std::string &name, const std::string &value)
+void emu_options::set_system_name(const std::string &new_system_name)
 {
-	if (name == OPTION_COIN_IMPULSE)
+	const game_driver *new_system = nullptr;
+
+	// was a system name specified?
+	if (!new_system_name.empty())
 	{
-		m_coin_impulse = int_value(OPTION_COIN_IMPULSE);
+		// if so, find it (and error if it cannot be found)
+		int index = driver_list::find(core_filename_extract_base(new_system_name, true).c_str());
```

Comment:
Another case of preserving existing logic; I thought that this was weird too.

In the current code (and MAME 0.184 and 0.185), I can do something like "mame64 pacman.foo", and it will load pacman.  This logic enables that behavior; if you want I can remove it.